src/solvers/refinement/string_constraint_generator_concat.cpp

```diff
@@ -44,28 +38,26 @@ exprt string_constraint_generatort::add_axioms_for_concat_substr(
   const exprt &start_index,
   const exprt &end_index)
 {
-  binary_relation_exprt prem(end_index, ID_gt, start_index);
-
-  exprt res_length=plus_exprt_with_overflow_check(
-    s1.length(), minus_exprt(end_index, start_index));
-  implies_exprt a1(prem, equal_exprt(res.length(), res_length));
+  const typet &index_type = start_index.type();
+  const exprt start1 = maximum(start_index, from_integer(0, index_type));
+  const exprt end1 = maximum(minimum(end_index, s2.length()), start1);
+  const plus_exprt res_length(s1.length(), minus_exprt(end1, start1));
+  const equal_exprt a1(res.length(), res_length);
   lemmas.push_back(a1);
```

Comment:
You could group those `a1`, `a2`, ... into:
```c++
lemmas.push_back([&] {
  exprt lemma;
  ...
  return lemma;
}());
```
That way you wouldn't have to worry about numbering these things.

Reply:
Numbering the axioms is not really a problem because they refer to these numbers in the documentation, but scoping some of these variables is definitely good, because I've made several times mistake where a universal symbol was used in two different constraints, so I'm going to follow your suggestion and add a comment to say to which axiom this corresponds